src/gui/qgsmaptoolcapture.cpp

```diff
@@ -289,10 +291,40 @@ bool QgsMapToolCapture::tracingAddVertex( const QgsPointXY &point )
     mSnappingMatches.append( QgsPointLocator::Match() );
   }
 
+  // If the layer supports curves, we de-approximate curves
+  QgsVectorLayer *vlayer = qobject_cast<QgsVectorLayer *>( mCanvas->currentLayer() );
+  if ( vlayer->dataProvider()->capabilities().testFlag( QgsVectorDataProvider::Capability::CircularGeometries ) )
+  {
+    // We convert the capture curve to OGR geometry
+    OGRGeometry *ogrCaptureCurve;
+    OGRGeometryFactory::createFromWkt( mCaptureCurve.asWkt().toLocal8Bit().data(), nullptr, &ogrCaptureCurve );
+
+    // We de-approximate the curves using OGR (the geometry must be forced to linear first, as a mCaptureCurve is a CompoundCurve)
+    OGRGeometry *ogrLinestring = ogrCaptureCurve->getLinearGeometry(); // it must be forced to linear, as captuecurve is a compound curve
+    OGRGeometry *ogrRecurved = ogrLinestring->getCurveGeometry();
+
+    // We save back to mCaptureCurve
+    mCaptureCurve.clear();
+    mCaptureCurve.fromWkt( QString::fromStdString( ogrRecurved->exportToWkt() ) );
```

Comment:
I wonder if we should encapsulate this code in qgsgeometryutils.h so it can be used from everywhere. And potentially replaced with a more stable version in one single place - or removed in one single place once things have been "properly fixed" in underlying libraries.

Reply:
Yes I wasn't really sure where to implement this... There's also `QgsAbstractGeometry.segmentize()`, so it would be make sense to add`QgsAbstractGeometry::desegmentize()` !

Reply:
Good point.
There's quite some overhead to it though at the moment. So I wonder to what extent we want to bring it "that deep into core" in its current shape. Very open to others opinions !